### PR TITLE
HttT General Improvements

### PR DESCRIPTION
**Pull request by GregoryLundberg**
This Pull Request is part of a series of corrections and improvements to the HttT ('Heir to the Throne') campaign.

When merging or testing, please be aware that a PR deeper in the hierarchy will have merge conflicts unless each higher PR has already been applied. If this presents a problem, contact me and I will re-factor the hierarchy as required. At present, the best options to contact me are either as comments on one of these pull requests, or by creating an Issue on my personal fork.

**Hierarchy**

    master
    |
    +-- GregoryLundberg:GL_HttT_general_improvements ( PR #730 )
        |
        +-- GregoryLundberg:GL_HttT_S05b_use_advisor ( PR #731 )
        |
        +-- GregoryLundberg:GL_HttT_S05b_randomize_temples ( PR #732 )
        |
        +-- GregoryLundberg:GL_HttT_S06_logical_thieves ( PR #733 )
        |
        +-- GregoryLundberg:GL_HttT_S07_ambushers_reduced ( PR #734 )
        |
        +-- GregoryLundberg:GL_HttT_S13_snow_detritus ( PR #735 )
        |
        +-- GregoryLundberg:GL_HttT_S20b_wose_assistance ( PR #736 )
        |
        +-- GregoryLundberg:GL_HttT_S22_gryphons_return ( PR #737 )

**GregoryLundberg:GL_HttT_general_improvements**
These are general improvements of my own. Each patch is fairly localized and should be easy to follow. Individual patches should be easy to cherry pick, if desired. The only complex patch is the last, which touches many files, and gets the debug command 'choose_level' working.

**GregoryLundberg:GL_HttT_S05b_use_advisor**
This implements a TODO item. It selects the advisor Konrad will have at the start of S06 ('The Siege of Elensefar') and uses that advisor for a comment, apparently coming from inside the ship, when it arrives.

**GregoryLundberg:GL_HttT_S05b_randomize_temples**
This implements a TODO item. It randomizes the temple contents.

**GregoryLundberg:GL_HttT_S06_logical_thieves**
This is an improvement of my own. It handles the Thieves' Guild in a more logical manner. Mainly, they will simply join forces in cases where Konrad has already passed the point where an offer of help makes sense.

**GregoryLundberg:GL_HttT_S07_ambushers_reduced**
This implements a TODO item. The number of ambushes reduces based upon the number of enemy leaders killed in S01 ('The Elves Besieged').

**GregoryLundberg:GL_HttT_S13_snow_detritus**
This implements a TODO item. It adds random snow on up to 1/3rd the non-snow tiles, based upon the snow coverage (that is, turns to completion) from S12 ('Northern Winter').

**GregoryLundberg:GL_HttT_S20b_wose_assistance**
This implements a TODO item. It adds a bonus sub-quest assisting the wose in dealing with the undead. The reward is a book which, when read, grants forest movement cost 1, forest defense 70%, and forest ambush (as with an Elvish Ranger). This reward was chosen because it offers some assistance in S22 ('Return to Wesnoth') and has limited, or no, use in the following scenarios.

**GregoryLundberg:GL_HttT_S22_gryphons_return**
This implements a TODO item. If Konrad did not injure or kill the gryphons in S10 ('Gryphon Mountain'), a flight of gryphons return to assist Konrad in S22 ('Return to Wesnoth'). Remember, for the gryphons to arrive, Konrad must have forgone their eggs and, therefore, has no Gryphon Riders. This lack of Gryphon Riders will be most felt in the final two scenario, and especially on the large, open plains of S23 ('The Test of the Clans').

**KNOWN BUGS**

Multiple-tile units with animation leave visual artifacts as they move. This is most noticeable with the Gryphons and Gryphon Riders as they move using move_unit_fake in S10 ('Gryphon Mountain'), S14 ('Plunging into Darkness') and S22 ('Return to Wesnoth'). The artifacts clear fairly quickly, and are all removed at the end of the movement.

The game engine appears to have memory-leak issues. When play-testing, after loading several dozen scenario, memory usage will have grown quite large. This can cause the system to begin to swap (if your system has swap space), which severely degrades performance. To work around the issue, I keep an eye on memory use and restart the engine when it approaches the limit.

There are a few remaining user-interface issues such as the popup messages sometimes not appearing.

**NOTES**

I have tested each fork to ensure no merge conflicts occur, provided the required upper-level forks have already been merged. I found no problems using simple merges. I did, however, run into problems with rebase. If you need to rebase after merging these forks and get merge conflicts, abort your rebase (git rebase --abort). In general, such merge conflicts occur because git can re-order the commits; but that re-ordering can be avoided, preventing merge conflicts, with an interactive rebase. Contact me if you need assistance using rebase.

The debug command choose_level works well, now. In general, when testing, you can switch from any scenario to any other. Internal state, such as which path you took from S04 ('The Bay of Pearls') .. S05a ('Muff Malal's Peninsula') or S05b ('The Isle of the Damned') .. or the amount of snow which fell in S12 ('Northern Winter') is maintained until you visit the scenario which sets the state. Thus, if you want to test with the various incarnations of Moremirmu in S09 ('The Valley of Death'), you simply need to use choose_level to jump to S05a or S05b, then jump back to S09 once you have the desired state.

The debug command next_level also works well. But, since there is no way to know if you used next_level command, it always advances to the default next level. At points where there is a branch, S04 ('The Bay of Pearls') and S18 ('A Choice Must Be Made'), you must either play through, or use the choose_level command if you do not want to follow the default choice.

When using the choose_level command, be aware that game state, such as your gold balance, and your recall list (including your Heros) is NOT CHANGED. This means, yes, it is possible to advance your Heros and jump to S01 ('The Elves Besieged') and have Konrad, Li'sar and Kalenz, along with a host of other L3 units, even Gryphon Riders or Gryphons, if you have always wanted to clear the board of those pesky orcs.

There are times when a Hero (generally Delfador or Li'sar) cannot be on the map, either for story purposes or because that Hero is already in that scenario. In these cases your Hero is saved and, if you are using choose_level to jump around, will be restored once you leave the area.

Similarly, there are times when a unit will join Konrad's forces. For example, Haldiel in S02 ('Blackwater Port'). If you already have Haldiel in your recall list and choose_level to jump to Blackwater Port, Haldiel will be removed from your recall list, and a new Haldiel will appear at the appropriate time. For those units which only appear depending upon your actions, they will only be removed from your recall list (or the map) should you take the action which causes their appearance.

**CHANGES**

_07AUG2016_

GL_unique_items has been merged to master.

Spelling and grammar corrections on this document and PR titles. Unfortunately, the typo in a branch name has to stay.

Spelling and grammar corrections on commit messages.

Removed underscores in custom event names 'home destroyed' and 'victory dance'.

Spelling and grammar corrections in commit 'HttT S20a Fix bug: El'rien might be dead'

Spelling and grammar corrections in commit 'HttT S20b Fix bug: Bona-Melodia may be dead'

Corrected filter for moving side in commit 'HttT S08 Adjust closing-in area'

Spelling and grammar corrections, and eliminated po comments in commit 'HttT S20b Wose assistance quest'

_08AUG2016_

Clean up [objectives] handling for commit 'HttT S10 Change objectives'

Clean up the hidden advisor to use a galleon unit instead of a shallow copy for commit 'HttT S05b Use an Advisor'

Moved lua inline, removing custom tag, for commit 'HttT S13 Add some random snow'

Clean up [objectives] handling for commit 'HttT S19c Fix bug: Wrong objectives'

_10AUG2016_

Sync to master v1.13.5+dev (6235e18-Clean) and re-tested

_11AUG2016_

Removed commit 'HttT S21 Fix bug: No save'

Sync to master v1.13.5+dev (cf4f488-Clean) and re-tested

Finally got commit 'HttT S05b Use an Advisor' working just the way I always envisioned

_12AUG2016_

Removed workaround commit for [object] issues on S08, S11 and S16.

Sync to master v1.13.5+dev (265e41d-Clean)

_12AUG2016_

GL_HttT_errors has been merged to master

Sync to master v1.13.5+dev (12f7107-Clean) and re-tested

_13AUG2016_

Cleanup since [hide_unit] now works for commit 'HttT S05b Use an Advisor'

Audited and corrected state variables for 'HttT Debug choose_level works'

Clean up [objectives] handling for commit 'HttT S20b Wose assistance quest'

Only leaders count for commit 'HttT S07 Ambushers reduced'

Changed the subject from you to the unit in commit 'HttT S19a Improve Flaming Sword'

Changed the subject from you to the unit in commit 'HttT S19b Improve Void Armor'

Sync to master v1.13.5+dev (3cc2d09-Clean) and re-tested

_19AUG2016_

Removed commit 'HttT S12 Added ambiance (fog)'

Updated commit 'HttT S19a Improve Flaming Sword' for typo and [message] big fixed

Updated commit 'HttT S14 Consistent objectives' for new {HAS_NO_TURN_LIMIT} macro

Sync to master v1.13.5+dev (7fab085-Clean) and re-tested

_30AUG2016_

Sync to master v1.13.5+dev (b24fdbc-Clean) and re-tested

_06SEP2016_

Updated commit 'HttT Debug choose_level works' removing macro RECALL_ELSE

Sync to master v1.13.5+dev (206096c-Clean) and re-tested

_10SEP2016_

Re-enabled Paladin to use Flaming Sword.

Switched to using unit for speaker when taking Void Armor to allow gender-specific messages.

Sync to master v1.13.5+dev (eb3e27b-Clean) (only tested for the changes)

_13SEP2016_

Removed [kill] before [unit] and cleaned up Li'sar changing sides for choose_level

Sync to master v1.13.5+dev (1af1932-Clean) and re-tested.

### CODE REVIEW

**Comment by CelticMinstrel** [comment]
Some of the commits here seem to also be in the Errors PR... is that because this one depends on that one?

File: data/campaigns/Heir_To_The_Throne/scenarios/08_The_Princess_of_Wesnoth.cfg [hunk 1]
```diff
@@ -321,11 +321,16 @@
 
     [event]
         name=moveto
-        [filter]
+        [filter_condition]
             side=1
```
Comment (CelticMinstrel) [inline data/campaigns/Heir_To_The_Throne/scenarios/08_The_Princess_of_Wesnoth.cfg]: This `side=1` no longer has any effect now that it's in `[filter_condition]` rather than `[filter]`.

File: data/campaigns/Heir_To_The_Throne/scenarios/08_The_Princess_of_Wesnoth.cfg [hunk 2]
```diff
@@ -321,11 +321,16 @@
 
     [event]
         name=moveto
-        [filter]
+        [filter_condition]
```
Comment (CelticMinstrel) [inline data/campaigns/Heir_To_The_Throne/scenarios/08_The_Princess_of_Wesnoth.cfg]: The `side=1` no longer has any effect now that it's in `[filter_condition]` rather than `[filter]`.

File: data/campaigns/Heir_To_The_Throne/scenarios/13_The_Dwarven_Doors.cfg [hunk 3]
```diff
@@ -236,13 +236,17 @@
...
         #
-        [role]
-            type=Footpad,Thug,Poacher,Outlaw,Trapper,Bandit,Fugitive,Ranger,Huntsman,Highwayman
-            role=Outlaw_Advisor
-        [/role]
         [recall]
-            role=Outlaw_Advisor
+            id=Delurin
         [/recall]
+        [role] # If Delurin was just recalled, this will select him.
```
Comment (CelticMinstrel) [inline data/campaigns/Heir_To_The_Throne/scenarios/13_The_Dwarven_Doors.cfg]: I noticed something like this in the Errors PR too, but... how do you know that it'll select Delurin if he was just recalled?

File: data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg [hunk 4]
```diff
@@ -88,6 +90,9 @@
                     [/gold_carryover]
 
                     [note]
+                        [show_if]
+                            {VARIABLE_CONDITIONAL moved_too_close boolean_equals false}
+                        [/show_if]
```
Comment (CelticMinstrel) [inline data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg]: I'm pretty sure this will have no effect, since any `[show_if]` is evaluated only when the `[objective]` tag is executed and not every time the objectives are shown. (It's not very intuitive, I'll admit, and perhaps it should be changed.)

File: data/campaigns/Heir_To_The_Throne/scenarios/13_The_Dwarven_Doors.cfg (hunk 3 above)
Comment (GregoryLundberg) [inline data/campaigns/Heir_To_The_Throne/scenarios/13_The_Dwarven_Doors.cfg]: Because [role] checks on-map before it checks the recall list and, at this point, in prestart, we know exactly who is on the map can can match the types.

File: data/campaigns/Heir_To_The_Throne/scenarios/19c_Cliffs_of_Thoria.cfg [hunk 5]
```diff
@@ -188,6 +189,7 @@
         side=4
         no_leader=yes
         controller=ai
+        hidden=yes
```
Comment (CelticMinstrel) [inline data/campaigns/Heir_To_The_Throne/scenarios/19c_Cliffs_of_Thoria.cfg]: Maybe the ally should be unhidden once revealed?

**Comment by GregoryLundberg** [comment]
Each of the branches has all the general-improvement and error commits.

It's because git and/or github do no support branching except from master. Works fine locally but that could be because merge is seeing the duplicated commit UUID and skipping.

**Comment by GregoryLundberg** [comment]
zookeeper wanted stuff separated and it look good locally. the duplicated commits as a surprise when it got to GitHub

File: data/campaigns/Heir_To_The_Throne/scenarios/13_The_Dwarven_Doors.cfg (hunk 3 above)
Comment (GregoryLundberg) [inline data/campaigns/Heir_To_The_Throne/scenarios/13_The_Dwarven_Doors.cfg]: No idea why this diff is suddenly marked outdated. nothing has changed

File: data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg (hunk 4 above)
Comment (GregoryLundberg) [inline data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg]: I fire the objectives action when it should change and use silent= to keep quiet about it. Or, I should .. it seemed to work when I tested but i may have misses a path through the code.

File: data/campaigns/Heir_To_The_Throne/scenarios/19c_Cliffs_of_Thoria.cfg (hunk 5 above)
Comment (GregoryLundberg) [inline data/campaigns/Heir_To_The_Throne/scenarios/19c_Cliffs_of_Thoria.cfg]: It's the sergeant and he'll join forces with Konrad when you move a unit next to him. I thought about it but there seemed little point for one unit which just stands (lies) there until you do something and, then, the side is empty and needs re-hiding.

File: data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg (hunk 4 above)
Comment (CelticMinstrel) [inline data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg]: That was the case in other similar uses, but this one is only in prestart.

File: data/campaigns/Heir_To_The_Throne/scenarios/19c_Cliffs_of_Thoria.cfg (hunk 5 above)
Comment (CelticMinstrel) [inline data/campaigns/Heir_To_The_Throne/scenarios/19c_Cliffs_of_Thoria.cfg]: Ah, yeah, if it's a unit that'll join you then there's no point.

**Comment by CelticMinstrel** [comment]
Hopefully the duplicated commits will "disappear" once the dependent branches are merged.

I skipped over the final commit since I don't really understand what's going on there, but for the most part this all looks good to me.

File: data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg (hunk 4 above)
Comment (GregoryLundberg) [inline data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg]: Will re-test and adjust if needed.

I'm considering, if changes are needed forgetting zookeeper's desire for separate PR for each feature and going back to an all-in-one. It's cleaner and a LOT less confusing.

**Comment by GregoryLundberg** [comment]
It merges fine. I may have to rebase and force a push to get the PRs to update accordingly. No great problem, just a bit of management.

I tested and if you merge errors, then general improvements, then any order you want for the features it works fine. Just don't rebase after you do unless you do interactive rebase. Each change is independent but the way git keeps then internally is time-sequenced and a normal rebase moves then around causing merge conflict.

**Comment by GregoryLundberg** [comment]
The funny thing is, if I do the PR against my fork, the feature PRs show just the single commit. It's crossing the line over to upstream/master which caused all the duplicated commits. Makes some sense, I guess. And I s'pose the Git Police will be after me for Criminal Branch Abuse, with Intent.

File: data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg (hunk 4 above)
Comment (CelticMinstrel) [inline data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg]: I suggest first focusing on the Errors PR and getting that merged before updating this one (ie, force-pushing). Then focus on this one and get it merged before updating the others. An all-in-one PR isn't a bad thing generally, but in this case the sheer number of changes is pretty overwhelming, so that's a good reason to keep them split up in my opinion.

Comment (GregoryLundberg) [inline data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg]: OK. That's sort of the process I'm hoping for, anyway.

It's all just having my head on straight when rebasing and pushing. Pencil and paper are your friend!

Comment (GregoryLundberg) [inline data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg]: Re-tested for the [objectives] question about [show_if]

The objectives do not display when the condition changes (as desired).

The dependent variable changes when Delfador makes his comment that the Sceptre is near.

If the user then asks for the objectives, the note has disappeared (as desired).

So, no, when using [show_if] you do not need to execute an [objectives] action.

Which causes me to re-think my other changes for this. I'm probably doing the [objectives] action way too often and can make better use of [show_if]. I will examine that issue tomorrow.

Comment (CelticMinstrel) [inline data/campaigns/Heir_To_The_Throne/scenarios/17_Scepter_of_Fire.cfg]: Given the implementation of `[objectives]` (in data/lua/wml/objectives.lua), that's kinda surprising... maybe I missed something important though. It's rather complicated.

**Comment by gfgtdf** [comment]
I wonder why you need that 'RECALL_ELSE .. [unit]' stuff, placinf a [unit] in [side] should already do exactly that: recalling a unit with the given id and create it form the config if there is no such unit on the recall list.

**Comment by GregoryLundberg** [comment]
I'll test it. It's been a month or two since I worked on it. Most likely I'll yank the last commit and not worry about choose_level done this way.

I pulled the similar from TSG because we've gotten [filter_recall] and [modify_unit] working and that's changing what it takes to manage units both in normal use and for choose_level.

**Comment by CelticMinstrel** [comment]
That seems kind of counterintuitive to me, honestly. Is it actually documented?

**Comment by gfgtdf** [comment]
Not sure, but afaik every mainline campaign relies on this behviour.

**Comment by GregoryLundberg** [comment]
I do know that if I create a unit on the map with the same ID as a unit in the recall list that the unit on the recall list is deleted. I do not know whether the XP (as an example) reset or come from the unit on the recall list. I can test that.

I do also know that if I first recall the unit to the map and then create a unit with the same ID I end up with two units having the same ID and things begin to go downhill from there.

**Comment by gfgtdf** [comment]
> I do know that if I create a unit on the map with the same ID as a unit in the recall list that the unit on the recall list is deleted


Hmm no it rather recalls the unit form the recall list instead of of creating a new one, in particular every attribute in [unit] except id=, placement=, and x,y= is ignored in this case.

**Comment by GregoryLundberg** [comment]
Removed macro RECALL_ELSE since the [unit] command does exactly what the macro did.

File: data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg [hunk 6]
```diff
@@ -222,7 +222,7 @@
                                     [message]
                                         speaker=narrator
                                         image="wesnoth-icon.png"
-                                        message= _ "You struggle to lift and don the heavy plate. Once worn, however, it is amazingly comfortable. You have increased resistance to all physical damage!"
+                                        message= _ "$unit.name| struggles to lift and don the heavy plate. Once worn, however, it is amazingly comfortable. $unit.name| feels an increased resistance to all physical damage!"
```
Comment (ln-zookeeper) [inline data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg]: Any particular reason why the unit wasn't made the speaker here, like it is with the flaming sword? I presume the rationale with the flaming sword was to make sure the unit gets highlighted, and that would apply here as well.

Also something to be aware of: changing the text to refer to "$unit.name" instead of "you" is fine, but it only works if the player can never get a unit which doesn't have a name. I don't think that possibility exists currently in the campaign, though, so this should be safe. And maybe you already considered that, I'm just saying in case it's relevant in the future with similar changes in some other campaign.

File: data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg [hunk 7]
```diff
@@ -93,17 +128,19 @@
                                 description= _ "This massive blade was created centuries ago by long-forgotten elvish forgemasters, who imbued the bluish steel with an inner magical fire. Tongues of flame dance on the surface, giving the metal a flawless mirrored finish."
                                 cannot_use_message= _ "Only the leader of an army can wield this sword!"
                                 [filter]
-                                    type=Fighter,Commander,Lord,Princess,Battle Princess,Elvish Captain,Elvish Hero,Elvish Marshal,Elvish Champion,Paladin,Elvish Lord,Elvish High Lord
+                                    type=Fighter,Commander,Lord,Princess,Battle Princess,Elvish Captain,Elvish Hero,Elvish Marshal,Elvish Champion,Elvish Lord,Elvish High Lord
```
Comment (ln-zookeeper) [inline data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg]: Actually, I think Paladin seems like an entirely reasonable choice. From this point on, you'll encounter almost exclusively arcane-resistant enemies, so the normal sword mostly sucks. Getting magical, +25% damage and switching the damage type from a liability to a neutral (and occasionally good) one doesn't seem bad at all.

File: data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg (hunk 6 above)
Comment (CelticMinstrel) [inline data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg]: Well, in that case you could still do something like `$unit.name?The $unit.type|`.

File: data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg (hunk 7 above)
Comment (GregoryLundberg) [inline data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg]: I removed Paladin because it didn't appear to have any effect upon the Paladin so seemed a waste. I'll re-test to be sure.

File: data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg (hunk 6 above)
Comment (GregoryLundberg) [inline data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg]: I thought about it and went with "least change". I would prefer the unit to speak as with the sword, but didn't want to make a large style change.

I will re-word to be more like the flaming sword.  "He" and "She" would work better in the latter phrase.

I can make a unit with no name and, yes, expect it to have a problem. But I don't think it's possible for the player, here, to have one without debug mode.

File: data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg (hunk 7 above)
Comment (CelticMinstrel) [inline data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg]: If you read a few lines lower in the `[effect]`, it specifically says `type=blade`, which means it indeed has no effect on the paladin's lance (which is pierce) or sword (which is arcane). If you want the paladin to be able to get it, you'll need to change the weapon filter in the `[effect]`, but I'm not sure there'd be much benefit to switching arcane to fire.

Comment (gfgtdf) [inline data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg]: I wonder whether it migth be better to filter on name (id) (would maybe need to be implemented in the c++ engine ), So then it could be some regex check like name=.*blade.*|.*sword.*|.*saber.* this would then also exlude other non-sword melee blade-type attacks like dwarwes axes.

Comment (CelticMinstrel) [inline data/campaigns/Heir_To_The_Throne/utils/httt_utils.cfg]: Weapon filters currently support name, type, and range as a comma-separated list of possible values. I'm not sure that filtering by name is a good idea though,

**Comment by GregoryLundberg** [comment]
I will sync-to-master, change the Void Armor to use gender-specific messages similar to how the Flaming Sword does it, and push an update.

On the issue with the Paladin not seeing and effect and being dropped from support for the Flaming Sword: I'm not sure what, if anything, to do about it. It does seem reasonable, especially seeing how Paladin is often treated as a sort-of-White-Mage elsewhere. So, I'm waiting for a clear idea of how to handle the issue.

**Comment by ln-zookeeper** [comment]
Paladin not receiving the effect is my fault, it seems (dcbf3c0f). So if weapon filter name now accepts a comma-separated list, then the problem should be fixed by simply using name=sword,saber as the filter.

**Comment by GregoryLundberg** [comment]
Tested Li'sar and she gets the effect from the Flaming Sword.

If there are no more issues, I believe this is now ready to merge.

**Comment by ln-zookeeper** [comment]
It looks like you're performing all sorts of weird killing/storing of units (Ulfdain, mermen, Li'sar), apparently in order to :cl-proof them so that if you already have a unit and :cl back to a scenario where it appears, you won't end up with two units with the same id? That's what I concluded after staring at the code for quite a while.

The code really looks weird and frankly nonsensical when you see it, so there should at least be comments to clarify what's going on and why. Not to mention that giving the player units with a unique id isn't exactly a rare occurrence. A hacky WML workaround to :cl-proof some hero units is fine if it's also commented, but to add one for what ought to be hundreds of units across all the campaigns seems like getting into crazy territory.

Didn't creating a [unit] result in a recall if a unit with that id already exists on your recall list? I thought I saw someone mention that, and if that's true then I don't see what the point of the killing is.

**Comment by CelticMinstrel** [comment]
There might still be an issue of "what if that unit had already been recalled", though. `[unit]` happily creates another unit with the same ID if one already exists on the map.

**Comment by GregoryLundberg** [comment]
Yes, it's in support of debug|change_level.

And I didn't understand how [unit] worked. But that was mainly with the (now gone) RECALL_ELSE macro. This code, however, pre-dates removal of that macro; so there may be some tweaks which would make it more clear, or reduce the need for it. (I presume the comment is about how I [kill] before [unit] for the non-hero characters we collect along the way.)

Celtic Minstrel is correct, [unit] does not kill a unit already on the map, but it does kill the unit in the recall list.

My concern is that the game remain rational, even in the face of debug|choose_level. And things go seriously wrong if two units have the same id.

The 'correct' solution would be to ensure that all id on the map are unique; but that's probably too large a change and should be discussed before starting work. I doubt it would be a breaking change, though, since things do go awry eventually and, one would hope, designers attempting it would test and correct.

**Comment by ln-zookeeper** [comment]
True, the problem would still exist in case the player recalled a unit they would encounter later in the scenario. However, it's one thing for a campaign to support using :cl to jump from any scenario to any other scenario without any obvious breakage (such as missing hero units), and quite another for it to try to account for the scenarios being actually played like that without any error states or inconsistencies occurring along the way. Sure, if we can have the latter at no cost then it's great, but in this case the cost is a baffling pile of very unintuitive undocumented [kill]s as well as the way Li'sar is getting stored/unstored everywhere.

Even for preserving heroes/leaders/etc when hopping back and forth between scenarios it's quite enough to make sure that the correct units exist where they should; they don't need to actually be the same units and retain their XP, for example. If you hop from S20 to S8 and then to S18, Li'sar can just revert back to a fresh scepterless lvl2 Princess with 0XP along the way and that shouldn't really be an inconvenience for anyone.

Basically, :cl is a feature only for debugging and testing and everyone using it should be very aware that they can't expect campaign logic to be able to take arbitrary use of it into account. Now, as said I very much support the idea of having the basics work, for example starting up HttT, immediately hopping to S19c and conveniently having Kalenz and Li'sar present, but I think even the idea of being able to safely :cl to an earlier scenario is kind of stretching it. Again, if it just works then that's nice, but it shouldn't be enabled by complicated, hard to understand and undocumented workarounds.

**Comment by Vultraz** [comment]
Just seconding everything @ln-zookeeper said.

**Comment by GregoryLundberg** [comment]
I will remove the commit for getting debug change_level and next_level working. I'd probably do it different it I started over on it, anyway.
